### PR TITLE
Make Get-ChildItem honor Depth parameter with Include/Exclude

### PR DESCRIPTION
**Pull request by Windos**
Addresses issue #3726 by adding decrementing depth to ProcessPathItems.

Includes tests specifically for the examples listed in the issue and also testing that recursion still works when excluding or including a string.

### CODE REVIEW

File: src/System.Management.Automation/engine/SessionStateContainer.cs [hunk 1]
```diff
@@ -1593,7 +1593,7 @@ internal Collection<PSObject> GetChildItems(string[] paths, bool recurse, uint d
                             }
 
                             int unUsedChildrenNotMatchingFilterCriteria = 0;
-                            ProcessPathItems(providerInstance, providerPath, recurse, context, out unUsedChildrenNotMatchingFilterCriteria, ProcessMode.Enumerate);
+                            ProcessPathItems(providerInstance, providerPath, recurse, context, out unUsedChildrenNotMatchingFilterCriteria, ProcessMode.Enumerate, depth: depth);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateContainer.cs]: I'd prefer to exclude parameters with defaults and use overloads.

Comment (Windos) [reply src/System.Management.Automation/engine/SessionStateContainer.cs]: Just pushed new commit swapping optional parameter for overload. Is it along the right track?

**Comment by iSazonov** [comment]
@daxian-dbw Can we ask @Windos to fix (replace with overloads) skipIsItemContainerCheck defaults in the PR?

**Comment by daxian-dbw** [comment]
Well, the optional parameter was introduced to be an alternative of overloads, to reduce the number of overloads of a method.
If we remove `skipIsItemContainerCheck` by adding overloads, then there will be 4 overloads of `ProcessPathItems`, which IMHO might be cumbersome.

The optional parameter should never be used for public APIs or even internal APIs that may be called from other friend assemblies because the default value is baked into the caller, and that means if we change the default value, the caller assembly needs to be re-compile to get the new default value.
However, it's fine for private methods, as long as it doesn't affect readability much.

**Comment by iSazonov** [comment]
@daxian-dbw I asked because adding second optional parameter reduced readability. Do we have to create overloads in this case or can we just replace its with one method with all the parameters?

**Comment by Windos** [comment]
I'm happy to put the work in regardless of which direction is decided on.

**Comment by daxian-dbw** [comment]
@iSazonov I think you made the right call. Adding the second optional parameter would worsen the readability. I think the current two overloads look good.

**Comment by adityapatwardhan** [comment]
@iSazonov @daxian-dbw can you update your review

**Comment by iSazonov** [comment]
@daxian-dbw Is `uint depth` parameter on best place?

**Review by adityapatwardhan — APPROVED** [review]
LGTM

**Comment by adityapatwardhan** [comment]
@daxian-dbw  Can you update your review?

**Review by daxian-dbw — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@Windos Thanks for your contribution!

**Pull request merged**